Post Title Block: add typography formatting options

creativecoder opened the pull request:
## Description

Adds text formatting options to the Post Title block using experimental block supports declarations.

### Summary of changes

- Adds font weight, font style, and text transform typography controls to the Post Title block.

Partially addresses #31117, but leaves out text decoration, for now

## How has this been tested?

- Install and activate the [tt1-blocks theme](https://github.com/WordPress/theme-experiments/tree/e80d8d5c59859fef449d83d428ddeca7906ad830/tt1-blocks)
- Open a template in the Site Editor that uses a Post Title block, such as Index or Single
- Test the additional typography settings for the block. Note that they apply to the entire block contents, they are _not_ inline controls
- Make sure they work as expected with "Make title a link" on and off

## Screenshots <!-- if applicable -->

<img width="280" alt="Screen Shot 2021-09-14 at 14 38 55" src="https://user-images.githubusercontent.com/1699996/133323132-44ba50b3-8970-432e-b913-3344f711ccd7.png">


## Types of changes

New feature (non-breaking change which adds functionality)

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

creativecoder committed:
Post Title Block: add typography formatting options

creativecoder commented on packages/block-editor/src/components/text-decoration-control/index.js:
```diff
@@ -2,10 +2,19 @@
+	{
+		name: __( 'None' ),
+		value: 'none',
+		icon: textColor,
```
I could use some suggestions about what icon to use here for indicating `text-decoration: none` cc @jasmussen @jameskoster

jasmussen replied:
Thanks for the ping. 

> Adds a "None" option to text decoration controls, for overriding the theme settings.

The case for "overriding theme settings" here is the common case of someting like `a { text-decoration: underline; }` correct? Or were you referring to some theme.json provided switch?

In any case, from what I can tell there already exists a "none" button:

<img width="591" alt="Screenshot 2021-05-10 at 08 08 46" src="https://user-images.githubusercontent.com/1204802/117613560-60e06a80-b167-11eb-9b3f-caa6ecb57e32.png">

<img width="106" alt="Screenshot 2021-05-10 at 08 08 28" src="https://user-images.githubusercontent.com/1204802/117613565-62119780-b167-11eb-8a83-51483f416c64.png">

The icon is not the great, and while "underline" is fine, the three of them together aren't harmoniously balanced to sit next to each other. I like what figma does here, having a generic "none" icon that is reused for every property that has a "none":

<img width="249" alt="Screenshot 2021-05-10 at 08 09 30" src="https://user-images.githubusercontent.com/1204802/117613650-7fdefc80-b167-11eb-90db-2d069e401a02.png">

However updated icons feels like a task for another day, and shouldn't block this effort.

oandregal replied:
If this bit is extracted to a different PR it can land faster (I'm not sure about the state of the other things I've mentioned at https://github.com/WordPress/gutenberg/pull/31623/files#r629277217 may or may not be ready to land but we need them for this).

creativecoder commented on packages/block-library/src/post-title/style.scss:
```diff
@@ -1,3 +1,7 @@
 .wp-block-post-title a {
 	display: inline-block;
+
+	// Inherit the styles from the main block element, so that text decoration settings
+	// for the block apply to post title links.
+	text-decoration: inherit;
```
This is basically a hack to apply the inline text-decoration styles to the child `<a>` element, when the post title uses a link. One undesired side-effect is that the default underlined browser link style is overridden (an accessibility concern).

Alternatives I can think of

- Update `edit.js` and `render_block_core_post_title` for this block to apply the inline styles to `<a>`, when present (rather than the parent heading element)
- Update the experimental style support to allow targeting an element in the block, (e.g. [how duotone works](https://github.com/WordPress/gutenberg/blob/b9232527214cd3d523badc4a66b529b7efe615ca/docs/reference-guides/block-api/block-supports.md#color__experimentalduotone))

creativecoder replied:
cc @nosolosw , as I believe the ability to specify styles that target elements within blocks is under active development.

oandregal replied:
At the block level (hooks), I think we'd want to use the `__experimentalSkipSerialization` flag. The post title block would be responsible for serializing the value directly to the `a` element. Context: https://github.com/WordPress/gutenberg/issues/28913 and this particular PR is necessary for this to work https://github.com/WordPress/gutenberg/pull/30880 

At the global level (via theme.json), we can already target `a` elements within the post title block:

```
styles: {
    core/post-title: {
        elements: {
            link: {
                typography: {
                    textDecoration: "value"
                }
            }
        }
    }
}
```

ntsekouras replied:
Since https://github.com/WordPress/gutenberg/pull/30880 has landed, should this PR be refreshed?

paaljoachim commented:
Thanks for working on this @creativecoder Grant!

I added the Post Title block to a page.

Before:
![Screen Shot 2021-05-09 at 23 41 43](https://user-images.githubusercontent.com/5323259/117587940-c2311b00-b120-11eb-81bd-ba97c5c1503a.png)

After:
![Screen Shot 2021-05-10 at 00 11 15](https://user-images.githubusercontent.com/5323259/117588551-4933c280-b124-11eb-9a5d-1e835deb3e91.png)

Thank you for the additional controls! I went through and tested all of these. The controls works nicely!

As I added the block to a page. The page end up with having two titles. The static title and the Post Title block title.
Related issue that is being worked on these days. To moving the title to the top bar + having somekind of placeholder mini static title for the default title. https://github.com/WordPress/gutenberg/issues/27093 
Could also use additional feedback and likely developer help. Thanks.

jasmussen commented:
Cool, thanks for your work here. This is before:

<img width="288" alt="Screenshot 2021-05-10 at 08 14 37" src="https://user-images.githubusercontent.com/1204802/117613948-e95f0b00-b167-11eb-9708-616fbf074be7.png">

This is after:

<img width="292" alt="Screenshot 2021-05-10 at 08 15 10" src="https://user-images.githubusercontent.com/1204802/117613966-eebc5580-b167-11eb-9840-6e35daec502e.png">

I know I've noted this in a couple of adjacent tickets, but it's important to reiterate. These text-decoration and letter case properties work well but are somewhat exotic/advanced controls, meaning until we have a better syste, for progressively adding these (as outlined in #27331), we should only be adding them to exotic/advanced blocks. In this case, adding it to Post Title seems fine, as that is exactly that: an advanced block as part of the FSE effort. So that's just to say, we need to refine the user experience before adding it to something like Paragraph.

Addison-Stavlo commented:
Testing this out everything (except the below) seems to work well.

Im not sure if this is an issue with the hook used or specific to its implementation here, but when I try to press the "A" button in text decoration it doesn't seem to change anything:

![Screen Shot 2021-05-10 at 10 54 41 AM](https://user-images.githubusercontent.com/28742426/117679576-65fce400-b17e-11eb-8732-f71821ffa54a.png)

Also (more of a critique of this tool in general), there is no tooltip when hovering these buttons.  I can't actually remember what "A" is supposed to do! 😆

paaljoachim commented:
@Addison-Stavlo
You are writing exactly the things that came to my mind which I forgot to write.  :-)
As I too did not notice any reaction when clicking the "A" button. So when I wondered what it was I tried hovering the cursor on the "A" and expected a tooltip to show up, but there was none.

jasmussen commented:
I had to [inspect the DOM](https://github.com/WordPress/gutenberg/pull/31623#discussion_r629072664) to find the Aria label that says "None" to figure out what the A meant, so the icon isn't super clear as "none". 

But digging a bit further, I'm now not sure where it's coming from. Looking at the Navigation block, which I believe is one of the first blocks to use the text-decoration control, it doesn't look like it has that "none" option:

<img width="248" alt="Screenshot 2021-05-11 at 09 29 58" src="https://user-images.githubusercontent.com/1204802/117776262-ac644880-b23b-11eb-880d-7493a3845456.png">

@aaronrobertshaw I think you worked on text decoration, do you recall where that's from?

aaronrobertshaw commented:
> @aaronrobertshaw I think you worked on text decoration, do you recall where that's from?

This isn't from the initial implementation of the text decoration block support. It was [added as part of this PR we are looking at](https://github.com/WordPress/gutenberg/pull/31623/files#diff-6857412d605d9ff09b16e6dfdfb9e974758f3f8c9f3ef2b02eebd266e6302058R13-R16).

My guess is that the `A` was to represent normal text, no underline, no strikethrough. The character `A` was used to match the text transform support icons. That of course could just me be trying to belatedly interpret reasoning.

Originally, it was to use a ButtonGroup ( https://github.com/WordPress/gutenberg/pull/26059#issuecomment-712623443 ) including an option for `none` for the user to be able to override something set by the theme. When it was decided it should use a segmented control, the `none` option wasn't included (https://github.com/WordPress/gutenberg/pull/26060#issuecomment-713396578).

jasmussen commented:
> My guess is that the A was to represent normal text, no underline, no strikethrough. The character A was used to match the text transform support icons. That of course could just me be trying to belated interpret reasoning.

Good assessment, and I think it could work if it shared more DNA with the underline and strikethrough icons, say had the same font size and throughline. 

> Originally, it was to use a ButtonGroup ( #26059 (comment) ) including an option for none for the user to be able to override something set by the theme. When it was decided it should use a segmented control, the none option wasn't included (#26060 (comment)).

Thanks for the links, the segmented control still makes sense since for this property you can still only actually choose one at a time, not multiple. Thinking in terms of the user experience, the tricky bit here is to help indicate why the "none" option exists at all, when you can just untoggle one of the others. I.e. what's the difference between explicitly toggling "none" and _not toggling any of the options_? — I know the technical difference, but it's hard to convey. 

We might just need better icons here and things will fall in place. I can look at this when I get a moment.

aaronrobertshaw commented:
>  I.e. what's the difference between explicitly toggling "none" and not toggling any of the options? — I know the technical difference, but it's hard to convey.

Exactly. The difference between an explicit none or "clear decoration" and the "default" or "inherited". It's a tricky one to convey via an icon.

jameskoster commented:
I like how Figma represents "none" with a dash:

<img width="257" alt="Screenshot 2021-05-11 at 10 11 35" src="https://user-images.githubusercontent.com/846565/117790507-51cdeb00-b241-11eb-90e2-8fcca11dde83.png">

This makes the pattern consistent and easy to identify across different controls.

Using a letter like the "A" is always going to be somewhat ambiguous since it can easily be confused for "upper case" or "lower case". I know these aren't text decoration properties but the average user probably does not, and the Letter Case control may not always be present to guide them.

creativecoder commented:
Thank you for all the feedback and sorry for the confusion on the "none" setting and the icon, let me try to clarify:

I grabbed the `textColor` icon as a placeholder to represent `text-decoration: none`, because it looked (kind of) like "unstyled" text. But I 100% agree that doesn't have enough intrinsic meaning, we need something else. Will try the dash, (please let me know the best source of an SVG for that 🙂).

The option to use `none` here is important for overriding theme and/or browser styles. For example, Post Titles in the index template are always underlined in TT1 Blocks, but this isn't actually because of the theme, it's the browser's default style for links. Without the `none` option, there's no way to turn off the underline.

I'll proceed with the following:

- Break out the addition of a "none" option for `__experimentalTextDecoration` into a separate PR as @nosolosw  suggested--and maybe look into the labeling issue, while I'm at it.
- In this PR, remove the css "hack" and try out the `__experimentalSkipSerialization` flag to target the text-decoration styles on the `a` tag. This should fix the issue of the None button seeming to do nothing, as @Addison-Stavlo and @paaljoachim pointed out.

Addison-Stavlo commented:
Now that I know that the regular "A" means "none"... this totally makes sense. 😆  🤦‍♀️ 

> The option to use none here is important for overriding theme and/or browser styles.

Great point!

carolinan commented:
A text that is not a link should almost never be underlined, this is a big accessibility issue.
I do not think it is suitable to expose the underline control to users.

alexstine commented:
I'll second @carolinan . My belief is most text formatting can get users in to trouble. Not so much for screen readers as much as it would effect people with low vision who rely on those different looking text clues.

I think this whole PR could use a review from the Accessibility team as a whole and see if any other of these formatting concerns could come up aside from the underline.

Thanks.

joedolson commented:
Underlining unlinked text does not directly trigger an accessibility problem, but for sighted users underlining text that is not linked creates usability problems. Users will reflexively expect the underlined text to be a link. I wouldn't consider this an accessibility problem, per se, since it impacts all users who are able to perceive the underline pretty much equally. 

Whether we should offer this as an option or not is somewhat a general question of whether WordPress should encourage best practices.

Underlined text that isn't linked is generally a poor practice from a design and usability standpoint. I generally think that WordPress should avoid offering this as an option. 

Regarding other options:

Does the strikethrough toggle the semantic state of the text? If it wraps the text in `<del>` elements, that's fine; that will be communicated to most modern screen readers. Text decoration only will not, however, and I can't think of any cases where strikethrough is used purely as decoration for text.

All caps can also create accessibility problems, but since users have always been able to get around that by typing in all caps, I don't see any reason not to provide that option.

creativecoder commented:
Separate PR for adding the "none" option to the text decoration controls: https://github.com/WordPress/gutenberg/pull/31768

alexstine commented:
A summary of the Accessibility teams discussion can be found here.

https://make.wordpress.org/accessibility/2021/05/14/accessibility-team-meeting-notes-may-14-2021/

ianstewart commented:
> It was decided that it would be better to remove the underline formatting option from the pull request for now and maybe revisit implementation later if there is a greater need.

It looks like removing the underline was the recommendation for a11y, is that correct?

carolinan commented:
The underline _itself,_ when the post title is linked, should be the default text decoration.

A formatting option should not make it possible to add underlines to the text when it is not a link.

At some point in the future, I trust that there will be ways to handle text decorations and hover styles and colors for links. It is unclear if is a good idea to expose it as controls to users.

jasmussen commented:
In my mind, exposing underline controls to users for links or headings make sense primarily in the sense that they would empower you to "fix" a bad theme that you otherwise love: maybe the theme comes with no underlines to those links, and by surfacing the tools you can add them back. You could also have done that by choosing another theme, or by writing custom CSS. But the underline tool itself is agnostic, it can be used for good or evil. 

What we can do is push towards good defaults and practices. Using [a progressive reveal system](https://github.com/WordPress/gutenberg/pull/32392), we can hide underline controls by default, adding friction. We can even show contextual warnings or notices when you use the features in a bad way, e.g. if you employ an underline on a paragraph, we can output a big text warning — just like the contrast checker — explaining why that's a bad idea. That kind of education, paired with the tools, might ultimately be the better solution than hiding the controls.

alaczek commented:
> In my mind, exposing underline controls to users for links or headings make sense primarily in the sense that they would empower you to "fix" a bad theme

Yes, totally agree. This has been my experience so far. Without these controls, there's simply no way to get around theme default choices.

How could we move this PR forward? From my perspective, the controls that would make the biggest impact design-wise are font weight and font style. If text decoration and transformation need more discussion, maybe we could break them out into a separate PR, and get the other enhancements shipped?

mtias commented:
Can we get all the other options except text-decoration merged and continue any approach there separately?

glendaviesnz commented:
@creativecoder are you wanting to continue with this PR or are you happy for someone else to make the change Matias suggests to get it over the line and merged?

creativecoder committed:
Revert text decoration changes

creativecoder commented:
@mtias @glendaviesnz Sorry for the delay, I lost track of this one. It's now updated and rebased, but omitting the text decoration controls.

A note about testing, there seems to be an issue with using the "Make a title a link" setting in the Post Title block, I get `Uncaught TypeError: can't access property "rendered", fullTitle is undefined` from `PostTitleEdit edit.js:97` when trying to toggle it. (cc @ntsekouras , in case https://github.com/WordPress/gutenberg/commit/7428d7e6a898b9232c01e8fe90a5b68ee6361191 is related). But that's not related to this change, as I see the same problem on `trunk`, as well.

This should be ready for review... there's not much to it, now!

apeatling reviewed (approved):
Tested and confirmed all settings are working as expected. :+1:

![2021-09-14 13 59 39](https://user-images.githubusercontent.com/1464705/133333200-23c00865-7c1b-44c3-ba04-3060ff79ef26.gif)

carolinan commented:
Works well for me, and I also do not see a JS error for the link (Chrome, Windows 10).

creativecoder commented:
> I also do not see a JS error for the link (Chrome, Windows 10)

Good to know! Maybe it's just me.

paaljoachim commented:
As the Post Title Block has been taken care of it seems natural to help @vdwijngaert Koen with the PR on **moving the Post/Page title to the top bar** so that we can perhaps get it into WP 5.9. 
The PR is here: https://github.com/WordPress/gutenberg/pull/31288
Thanks!
